minidns-dnssec/src/main/java/de/measite/minidns/dnssec/DNSSECClient.java

```diff
@@ -0,0 +1,525 @@
+                        if (unverifiedReasons.isEmpty()) {
+                            for (Record dsRecord : dss) {
+                                knownDelegations.put(dsRecord.name, (DS) dsRecord.payloadData);
+                            }
+                        }
+                    } catch (DNSSECValidationFailedException ignored) {
+                        // Not actually a problem, just an incomplete hint.
+                    }
+                }
+                result = null;
+            }
+            return createDnssecMessage(dnsMessage, result);
+        }
+        return null;
+    }
+
+    private DNSSECMessage createDnssecMessage(DNSMessage dnsMessage, Set<UnverifiedReason> result) {
+        Record[] answers = dnsMessage.getAnswers();
+        Record[] nameserverRecords = dnsMessage.getNameserverRecords();
+        Record[] additionalResourceRecords = dnsMessage.getAdditionalResourceRecords();
+        Set<Record> signatures = new HashSet<>();
+        extractSignatureRecords(signatures, answers);
+        extractSignatureRecords(signatures, nameserverRecords);
+        extractSignatureRecords(signatures, additionalResourceRecords);
+        if (stripSignatureRecords) {
+            answers = stripSignatureRecords(answers);
+            nameserverRecords = stripSignatureRecords(nameserverRecords);
+            additionalResourceRecords = stripSignatureRecords(additionalResourceRecords);
+        }
+        return new DNSSECMessage(dnsMessage, answers, nameserverRecords, additionalResourceRecords, signatures, result);
+    }
+
+    private void extractSignatureRecords(Set<Record> signatures, Record[] records) {
+        for (Record record : records) {
+            if (record.type == TYPE.RRSIG)
+                signatures.add(record);
+        }
+    }
+
+    private Record[] stripSignatureRecords(Record[] records) {
+        if (records.length == 0) return records;
+        List<Record> recordList = new ArrayList<>();
+        for (Record record : records) {
+            if (record.type != TYPE.RRSIG) {
+                recordList.add(record);
+            }
+        }
+        return recordList.toArray(new Record[recordList.size()]);
+    }
+
+    @Override
+    protected boolean isResponseCacheable(Question q, DNSMessage dnsMessage) {
+        dnsMessage.setAuthenticData(false); // At this state, a DNSMessage is never authentic!
+        return super.isResponseCacheable(q, dnsMessage);
+    }
+
+    private Set<UnverifiedReason> verify(DNSMessage dnsMessage) {
+        if (dnsMessage.getAnswers().length > 0) {
+            return verifyAnswer(dnsMessage);
+        } else {
+            return verifyNsec(dnsMessage);
+        }
+    }
+
+    private Set<UnverifiedReason> verifyAnswer(DNSMessage dnsMessage) {
+        Question q = dnsMessage.getQuestions()[0];
+        Record[] answers = dnsMessage.getAnswers();
+        List<Record> toBeVerified = new ArrayList<>(Arrays.asList(answers));
```

Comment:
I'm not sure if passing toBeVerified around is good. Would it be possible to split the dnsMessags answers right away?
- DNSKEY with SEP are special
- Everything else gets grouped by type (e.g. HashMap<String, Record[]>)
This should be a good input for e.g. verifySignatures. Some records could also get extra checks (NSEC/NSEC3) and thus unify the verify entry points. Do you think this would make sense?